modules/fvad/fvad.c

```diff
@@ -0,0 +1,365 @@
+	}
+
+	st->au = au;
+
+	list_append(&vad_txl, &st->le, st);
+
+	*stp = (struct aufilt_enc_st *)st;
+
+	return 0;
+}
+
+
+static int decode_update(struct aufilt_dec_st **stp, void **ctx,
+			 const struct aufilt *af, struct aufilt_prm *prm,
+			 const struct audio *au)
+{
+	struct vad_dec *st;
+	(void)ctx;
+
+	if (!stp || !af || !prm)
+		return EINVAL;
+
+	if (*stp)
+		return 0;
+
+	if (prm->ch != 1)
+		return EINVAL;
+
+	st = mem_zalloc(sizeof(*st), dec_destructor);
+	if (!st)
+		return ENOMEM;
+
+	st->fvad = fvad_new();
+	if (!st->fvad) {
+		mem_deref(st);
+		return ENOMEM;
+	}
+
+	int err = fvad_set_sample_rate(st->fvad, prm->srate);
+	if (err < 0) {
+		mem_deref(st);
+		return EINVAL;
+	}
+
+	st->au = au;
+
+	list_append(&vad_rxl, &st->le, st);
+
+	*stp = (struct aufilt_dec_st *)st;
+
+	return 0;
+}
+
+
+static bool auframe_vad(Fvad *fvad, struct auframe *af)
+{
+	static int chunk_times_ms[] = { 30, 20, 10 };
+
+	int16_t *buf = NULL;
+	int16_t *allocated = NULL;
+	bool detected = false;
+
+	/* convert to 16 bit linear */
+	if (af->fmt == AUFMT_S16LE) {
+		buf = af->sampv;
+	}
+	else {
+		allocated = (int16_t*)mem_alloc(sizeof(int16_t) * af->sampc,
+			NULL);
+		buf = allocated;
+		auconv_to_s16(buf, af->fmt, af->sampv, af->sampc);
+	}
+
+	size_t ms = af->sampc * 1000 / af->srate;
+	size_t pos = 0;
+
+	/* process all chunk_sizes that fvad accepts */
+	for (size_t chunk_time_index = 0;
+		chunk_time_index < RE_ARRAY_SIZE(chunk_times_ms);
+		++chunk_time_index) {
+
+		const size_t chunk_time = chunk_times_ms[chunk_time_index];
+		const size_t sampc = af->srate / 1000 * chunk_time;
+
+		for (size_t i = 0; i < ms / chunk_time; ++i) {
+			int err = fvad_process(fvad, buf + pos, sampc);
+			pos += sampc;
+			ms -= chunk_time;
+			if (err > 0) {
+				detected = true;
+				goto out;
+			}
+			else if (err < 0) {
+				warning("fvad_process(%d) failed", sampc);
+				goto out;
+			}
+		}
+	}
+
+	if (pos != af->sampc) {
+		warning("fvad_process: samples left over: %d",
+			af->sampc - pos);
+	}
+
+out:
+	mem_deref(allocated);
+
+	return detected;
+}
+
+
+static int encode(struct aufilt_enc_st *st, struct auframe *af)
+{
+	struct vad_enc *vad = (void *)st;
+
+	if (!st || !af)
+		return EINVAL;
+
+	bool vad_tx = auframe_vad(vad->fvad, af);
+
+	if (vad_tx != vad->vad_tx) {
+		vad->vad_tx = vad_tx;
+
+		if (vad_stderr)
+			print_vad(61, 32, false, vad_tx);
+
+		module_event("fvad", "vad", NULL, (struct call*)vad->call,
+			"%d", vad_tx);
+	}
+
+	return 0;
+}
+
+
+static int decode(struct aufilt_dec_st *st, struct auframe *af)
+{
+	struct vad_dec *vad = (void *)st;
+
+	if (!st || !af)
+		return EINVAL;
+
+	bool vad_rx = auframe_vad(vad->fvad, af);
+
+	if (vad_rx != vad->vad_rx) {
+		vad->vad_rx = vad_rx;
+
+		if (vad_stderr)
+			print_vad(64, 32, false, vad_rx);
+
+		module_event("fvad", "vad", NULL, (struct call*)vad->call,
+			"%d", vad_rx);
+	}
+
+	return 0;
+}
+
+/* To be used with uag_filter_calls */
+bool find_vad_rx(const struct call *call, void *arg)
```

Comment:
Should be static.

Use `uag_filter_calls()` in `decode_update()` and `encode_update()`.  You need a match handler and a list handler, which you can share between decode and encode.

The lists `vad_rxl` and `vad_txl` are not needed.

As I already wrote. You don't need to remember pointer `au` to audio object, only the `call`.

Reply:
Doh. I get it now. I'd misunderstood `uag_filter_calls`. It's fixed now.